### PR TITLE
feat(route53): `RecordSet` routing policy properties, `HeathCheck`

### LINKED ISSUE #4391: support routing policy setting for ARecord class
Please support routing policy parameters for ARecord in CDK.

Not all parameters are supported in CDK Route53 module currently. https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-recordset.html

any plan to support those in the future?

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by nmussy**
* `RecordSet` routing policy properties
* `HealthCheck` constructs
  * `EndpointHealthCheck` for IP and DNS health checks
  * `AlarmHealthCheck` for CloudWatch alarm health checks
  * `CalculatedHealthCheck` for health checks checking the health of other health checks
* export existing IP validation regular expression from `aws-ec2`

TODO: 
* [ ] Fix `awslint` errors due to base `HealthCheck` class
* [ ] Create higher level record generation methods? ([see comment](https://github.com/aws/aws-cdk/pull/4413#issuecomment-539565197))
* [ ] Fix `HealthCheck` tag generation
* [ ] Clarify if `region: Aws.REGION` is sufficient for `AlarmHealthCheck` CloudWatch alarms
* [ ] Health check integration tests?

Fixes #4391

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nmussy** [comment]
It might be interesting to add static methods (one per routing policy type) to build multiple records for the same `setIdentifier` (which could be auto-generated?). Something like:

```ts
public static generateFailoverRecords(primary: RecordSetOptions, secondary: RecordSetOptions)
public static generateWeightedRecords(...weightedRecords: RecordSetOptions & {weight: number})
// ...
```

File: packages/@aws-cdk/aws-route53/lib/heath-check/alarm-health-check.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+export class AlarmHealthCheck extends HealthCheck {
+    public constructor(scope: Construct, id: string, props: AlarmHealthCheckProps) {
+        const { alarm, ...baseProps } = props;
+        super(scope, id, {
+            type: AlarmHealthCheckType.CALCULATED,
+            ...baseProps,
+            alarmIdentifier: {
+                name: props.alarm.alarmName,
+                // FIXME not always true?
+                region: Aws.REGION,
```
Comment (nmussy) [inline packages/@aws-cdk/aws-route53/lib/heath-check/alarm-health-check.ts]: I don't know how to retrieve the alarm's region there

File: packages/@aws-cdk/aws-route53/lib/heath-check/health-check.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    public readonly healthCheckId: string;
+
+    protected constructor(scope: Construct, id: string, props: HealthCheckProps) {
+        super(scope, id);
+
+        // TODO tags?
+        const healthCheck = new CfnHealthCheck(this, 'Resource', { healthCheckConfig: props });
+        this.healthCheckId = healthCheck.ref;
+    }
+}
```
Comment (nmussy) [inline packages/@aws-cdk/aws-route53/lib/heath-check/health-check.ts]: awslint really doesn't like what I'm doing there. What would be the expected pattern?

File: packages/@aws-cdk/aws-route53/lib/heath-check/alarm-health-check.ts (hunk 1 above)
Comment (nmussy) [reply packages/@aws-cdk/aws-route53/lib/heath-check/alarm-health-check.ts]: I tried writing the following test:

```ts
// GIVEN
const app = new App();
const stack1 = new Stack(app, 'Stack1', {env: {region: 'us-east-1'}});
const stack2 = new Stack(app, 'Stack2', {env: {region: 'eu-west-1'}});

const alarm = new cloudwatch.Alarm(stack1, 'Alarm', {
  metric: new cloudwatch.Metric({
    metricName: 'Errors',
    namespace: 'my.namespace',
  }),
  threshold: 1,
  evaluationPeriods: 1,
});

// WHEN
new route53.AlarmHealthCheck(stack2, 'HealthCheck', {alarm});
```

But I'm getting the following error:
`Resolution error: Resolution error: Resolution error: Cannot use resource 'Stack1/Alarm' in a cross-environment fashion, the resource's physical name must be explicit set or use `PhysicalName.GENERATE_IF_NEEDED`.`

It this expected?

**Comment by shivlaks** [comment]
hey @nmussy - will start taking a look at this one tomorrow. Any high level feedback you'd like specifically while you work through your TODO list?

It might also help to separate the routing policy change (which addresses the issue referenced) and the HealthCheck constructs in separate PRs. but up to you!

**Comment by nmussy** [comment]
Hey @shivlaks,

That "todo list" should really be a "help me list" 😄. The biggest design decision is whether or not to add [higher level L2 methods](https://github.com/aws/aws-cdk/pull/4413#issuecomment-539565197) (which I think we should).

As for splitting the PR, I agree, it feels a little unnecessary to throw `HealthCheck` in there now, `HealthCheck` isn't required by the routing polices yet. However, if we do add those `generateX` methods, some of them might require a health check (e.g. `generateFailoverRecords`). The new health check PR would then block this one. So it's up to you, I don't mind either way.

**Comment by geekybaiyi** [comment]
any idea when this PR will be merged? 
thx!

**Comment by shivlaks** [comment]
closing this PR as it's stale to reflect that this task is available to pick up on the issue. PR can be reopened or a new one can be published in it's place.

**Pull request closed without merging**